File: backend/app/models/course.py
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, Text, UniqueConstraint, func
from app.database import Base


class Course(Base):
    __tablename__ = "courses"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(200), nullable=False)
    image_url = Column(String(500), nullable=True)
    current_assignment = Column(Integer, default=0)
    total_assignments = Column(Integer, default=0)
    next_exam = Column(Integer, nullable=True)
    total_exams = Column(Integer, default=0)
    status = Column(String(50), default="Not started")
    user_id = Column(Integer, ForeignKey("users.id"))
    credits = Column(Integer, default=3)
    google_id = Column(String(100), nullable=True)
    # SyllabusAI G12: optional link to a catalog subject (enrichment only).
    curriculum_subject_id = Column(Integer, ForeignKey("curriculum_subjects.id"), nullable=True)
    # Origin of the course row: manual | classroom | kb_tag | kb_folder.
    source_type = Column(String(20), default="manual")
    # Set when derived from a Second Brain `course:*` tag (Course ↔ KbTag).
    kb_tag_id = Column(Integer, ForeignKey("kb_tags.id"), nullable=True)
    # Set when derived from a top-level vault folder: which source the folder
    # lives in and the folder's relative path (e.g. "Operating Systems").
    kb_source_id = Column(Integer, ForeignKey("kb_sources.id"), nullable=True)
    kb_folder_path = Column(String(500), nullable=True)
    # Google Classroom course page (alternateLink) for classroom-sourced rows.
    classroom_url = Column(String(1000), nullable=True)
    # Number of Second Brain documents linked to the `course:*` tag or the
    # vault folder (distinct from total_assignments, which counts real
    # Assignment rows).
    kb_document_count = Column(Integer, default=0)
    # Folder-derived subjects: description / status / color read from the
    # folder's index.md frontmatter (description, status, color keys).
    description = Column(Text, nullable=True)
    color = Column(String(50), nullable=True)

    # Idempotency guarantees for KB-derived courses: one Course per tag (or
    # per source+folder) per owner. SQLite treats NULLs as distinct, so rows
    # of other origins never collide with these constraints.
    __table_args__ = (
        UniqueConstraint("user_id", "kb_tag_id", name="uq_courses_user_kb_tag"),
        UniqueConstraint("user_id", "kb_source_id", "kb_folder_path", name="uq_courses_user_kb_folder"),
    )


class CourseSyncLog(Base):
    """Latest per-user course-sync run (KB tag derivation diagnostics).

    One row per user, upserted on each run so the UI can explain the last
    sync (counts, sources/documents/tags seen, and any partial failures).
    """

    __tablename__ = "course_sync_log"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True, unique=True)
    created = Column(Integer, default=0)
    updated = Column(Integer, default=0)
    removed = Column(Integer, default=0)
    courses = Column(Integer, default=0)
    sources = Column(Integer, default=0)
    documents = Column(Integer, default=0)
    course_tags = Column(Integer, default=0)
    course_folders = Column(Integer, default=0)
    errors_json = Column(Text, nullable=True)
    synced_at = Column(DateTime, server_default=func.now())


class CourseGapAnalysis(Base):
    """Saved Gap Analysis for one course (one row per user + course).

    Gap Analysis is expensive relative to the page load, and the results are
    meant to persist: a subject's analysis is computed once, stored here, and
    reused on subsequent visits. It is only recomputed when the user explicitly
    re-runs it (``POST /api/courses/{id}/gaps/analyze``) or after a resync
    (the underlying data may have changed, so the cached copy is invalidated).
    """

    __tablename__ = "course_gap_analysis"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    course_id = Column(Integer, ForeignKey("courses.id"), nullable=False, index=True)
    # Full ``course_gaps`` payload (JSON-encoded).
    payload_json = Column(Text, nullable=False)
    analyzed_at = Column(DateTime, server_default=func.now())

    __table_args__ = (
        UniqueConstraint("user_id", "course_id", name="uq_course_gap_analysis_user_course"),
    )

File: frontend/src/services/api/types/course.ts
// Course / Subject Details page types (Second Brain + Classroom).
// Extracted from services/api.ts (F8 split).

import type { GapAnalysisResponse, GapCoverage, GapItem, GapPathPhase } from './gap'
import type { KbGraphResponse } from './kb'
import type { KbConceptGap } from './subject'

export interface Course {
  id: number; title: string; image_url: string | null;
  current_assignment: number; total_assignments: number;
  next_exam: number | null; total_exams: number;
  status: string; user_id: number;
  credits: number;
  progress_percentage?: number;
  kb_document_count?: number;
  // SyllabusAI G12 (Phase 78): optional link to a catalog subject.
  curriculum_subject_id?: number | null;
  // Origin of the row: manual | classroom | kb_tag | kb_folder.
  source_type?: string | null;
  kb_tag_id?: number | null;
  // Set when derived from a top-level vault folder (source + folder path).
  kb_source_id?: number | null;
  kb_folder_path?: string | null;
  // Folder-derived subjects: metadata read from the folder's index.md
  // frontmatter (description / status / color keys).
  description?: string | null;
  color?: string | null;
  google_id?: string | null;
  classroom_url?: string | null;
}

export interface CourseDocument {
  id: number;
  title: string;
  doc_type: string;
  path_rel: string | null;
  char_count: number;
  outline: { level: number; text: string; char_start: number }[] | null;
  tags: string[] | null;
  wikilinks: string[] | null;
  quality_score: number | null;
  reading_time_seconds: number | null;
  author: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface CourseSyncResult {
  created: number;
  updated: number;
  removed: number;
  courses: number;
  // Always returned by the backend derivation summary.
  sources: number;
  documents: number;
  course_tags: number;
  course_folders: number;
  errors: string[];
  synced_at: string;
}

export interface CourseSyncLog {
  created: number;
  updated: number;
  removed: number;
  courses: number;
  sources: number;
  documents: number;
  course_tags: number;
  course_folders: number;
  errors: string[];
  synced_at: string | null;
}

export interface CourseSyncStatus {
  sources: number;
  documents: number;
  course_tags: number;
  course_folders: number;
  last_sync: CourseSyncLog | null;
  google: { configured: boolean; connected: boolean; email: string | null };
}

/** One nested topic/subtopic with the documents that contain its heading. */
export interface CourseTopicNode {
  id: string;
  name: string;
  level: number;
  // folder → this topic is a vault folder under the subject (folder-derived
  // subjects only; the UI shows a folder icon + path breadcrumb); heading →
  // built from document headings. Absent on older payloads → treat as heading.
  origin?: 'folder' | 'heading';
  documents: CourseDocument[];
  children: CourseTopicNode[];
}

/** A concept mentioned by the subject's documents (related concepts). */
export interface CourseConcept {
  concept_id: number;
  name: string;
  definition: string | null;
  mentions: number;
  document_count: number;
  sources: { document_id: number; title: string }[];
}

/** A Classroom-sourced assignment for this course. */
export interface CourseClassroomAssignment {
  id: number;
  title: string;
  description: string | null;
  status: string;
  due_date: string | null;
}

/** One canonical folder/domain under a course (Second Brain folder hierarchy). */
export interface KbDomainNode {
  id: number;
  name: string;
  path: string;
  depth: number;
  doc_count: number;
  description: string | null;
  status: string | null;
  color: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  // Present on tree payloads (domain page / course domains grid).
  children?: KbDomainNode[];
}

export interface KbDomainListResponse {
  items: KbDomainNode[];
}

export interface KbDomainDetail {
  id: number;
  name: string;
  path: string;
  depth: number;
  doc_count: number;
  description: string | null;
  status: string | null;
  color: string | null;
  course: { id: number; title: string } | null;
  breadcrumb: { id: number | null; name: string; path: string }[];
  documents: {
    id: number;
    title: string;
    doc_type: string;
    path_rel: string | null;
    char_count: number;
    quality_score: number | null;
    reading_time_seconds: number | null;
    author: string | null;
    created_at: string | null;
    updated_at: string | null;
  }[];
  subfolders: KbDomainNode[];
}

/** Domain-scoped gap analysis — same actionable engine shape as course gaps. */
export interface KbDomainGapsResponse {
  summary?: GapAnalysisResponse['summary'] | null;
  strengths?: GapItem[];
  gaps?: GapItem[];
  path?: GapPathPhase[];
  next?: GapItem | null;
  domain?: string | null;
  coverage?: GapCoverage;
  document_count: number;
  folder: { id: number; name: string; path: string };
  // Saved-analysis metadata: cached=true means reused, not recomputed.
  cached?: boolean;
  analyzed_at?: string | null;
  new_notes_since_analysis?: number;
}

export interface CourseContentResponse {
  course: Course;
  second_brain: {
    documents: CourseDocument[];
    topics: CourseTopicNode[];
    // Canonical folder-derived domains (the Second Brain hierarchy).
    domains?: KbDomainNode[];
    unorganized_documents: CourseDocument[];
    concepts: CourseConcept[];
    graph: KbGraphResponse;
    document_count: number;
  };
  classroom: {
    linked: boolean;
    google_id: string | null;
    course_url: string | null;
    assignments: CourseClassroomAssignment[];
    total_assignments: number;
  };
}

/** Topic coverage gap reported by the subject gap analysis. */
export interface CourseTopicGap {
  topic: string;
  normalized: string;
  coverage: number;
  documents: number;
  is_gap: boolean;
}

export interface CourseGapsResponse {
  topics: CourseTopicGap[];
  concepts: KbConceptGap[];
  classroom: { assignments: CourseClassroomAssignment[]; total: number; pending: number };
  threshold: number;
  document_count: number;
  // Redesigned actionable gap engine (absent only on legacy backends).
  summary?: GapAnalysisResponse['summary'] | null;
  strengths?: GapItem[];
  gaps?: GapItem[];
  path?: GapPathPhase[];
  next?: GapItem | null;
  domain?: string | null;
  coverage?: GapCoverage;
  // Saved-analysis metadata: cached=true means this is the previously stored
  // result (reused, not recomputed); analyzed_at is when it was computed.
  cached?: boolean;
  analyzed_at?: string | null;
  // Staleness: how many Second Brain documents were added since this analysis
  // was computed (present on cached loads; 0 when nothing changed since).
  new_notes_since_analysis?: number;
}

export interface CourseResyncResult {
  kb: CourseSyncResult;
  classroom: {
    courses?: number;
    assignments?: number;
    source?: string;
    warning?: string;
    error?: string;
  } | null;
  course_removed?: boolean;
  content: CourseContentResponse | null;
}

export interface AcademicResource {
  id: string;
  title: string;
  type: string;
  /** Authoritative document type from the vault (defect #26) — drives the icon. */
  doc_type?: string | null;
  url: string | null;
  description: string | null;
}
